pkg/webhook/core.oam.dev/v1alpha2/application/helper.go

```diff
@@ -0,0 +1,51 @@
+package application
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/klog"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha2"
+)
+
+const (
+	errFmtTraitNameConflict = "traits name (%q) conflict between of component %q"
+)
+
+// ValidatingApp is used for validating Application
+type ValidatingApp struct {
+	app *v1alpha2.Application
+}
+
+// AppValidator provides functions to validate Application
+type AppValidator interface {
+	Validate(context.Context, ValidatingApp) []error
+}
+
+// AppValidateFunc implements function to validate Application
+type AppValidateFunc func(context.Context, ValidatingApp) []error
+
+// Validate validates Application
+func (fn AppValidateFunc) Validate(ctx context.Context, v ValidatingApp) []error {
+	return fn(ctx, v)
+}
+
+// ValidateTraitNameFn validates traitName should be unique in application
+func ValidateTraitNameFn(_ context.Context, v ValidatingApp) []error {
+	klog.Info("validate trait name conflicts ", "app name:", v.app.Name)
+	allErrs := make([]error, 0)
+	for _, component := range v.app.Spec.Components {
+		allTraitName := make(map[string]bool)
+		for _, trait := range component.Traits {
+			if _, ok := allTraitName[trait.Name]; ok {
+				allErrs = append(allErrs, field.Invalid(field.NewPath("spec"), v.app.Spec,
+					fmt.Sprintf(errFmtTraitNameConflict, trait.Name, component.Name)))
+			} else {
+				allTraitName[trait.Name] = true
+			}
```

Comment:
We should check the trait resource name not the trait name:

In the below example, the trait resource name is `service` and `ingress`. So you should parse the CUE template in this validation.

```
apiVersion: core.oam.dev/v1alpha2
kind: TraitDefinition
metadata:
  name: ingress
spec:
  extension:
    template: |
      parameter: {
        domain: string
        path: string
        exposePort: int
      }
      // trait template can have multiple outputs in one trait
      outputs: service: {
        apiVersion: "v1"
        kind: "Service"
        spec: {
          selector:
            app: context.name
          ports: [{
              port: parameter.exposePort
              targetPort: context.output.spec.template.spec.containers[0].ports[0].containerPort
            }]
        }
      }
      outputs: ingress: {
          apiVersion: "networking.k8s.io/v1beta1"
          kind: "Ingress"
          metadata:
            name: context.name
            labels: config: context.outputs.gameconfig.data.enemies
          spec: {
            rules: [{
              host: parameter.domain
              http: {
                paths: [{
                    path: parameter.path
                    backend: {
                      serviceName: context.name
                      servicePort: parameter.exposePort
                    }
                }]
              }
            }]
          }
      }
```